src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -2322,67 +2328,259 @@ internal override Binder GetBinder(SyntaxNode node)
 
             public override BoundStatement BindStatement(StatementSyntax node, DiagnosticBag diagnostics)
             {
+                BoundNode boundNode;
+
                 // Check the bound node cache to see if the statement was already bound.
-                BoundStatement synthesizedStatement = _semanticModel.GuardedGetSynthesizedStatementFromMap(node);
+                if (node.SyntaxTree == _semanticModel.SyntaxTree)
+                {
+                    BoundStatement synthesizedStatement = _semanticModel.GuardedGetSynthesizedStatementFromMap(node);
+
+                    if (synthesizedStatement != null)
+                    {
+                        return synthesizedStatement;
+                    }
+
+                    boundNode = TryGetBoundNodeFromMap(node);
 
-                if (synthesizedStatement != null)
+                    if (boundNode != null)
+                    {
+                        return (BoundStatement)boundNode;
+                    }
+                }
+
+                BoundStatement statement;
+                ref WeakReference<BoundNode> refToWeakReference = ref TryGetNodeFromSimpleProgramCache(node, out statement, out bool refToWeakReferenceIsValid);
+
+                if (statement is null)
                 {
-                    return synthesizedStatement;
+                    statement = TryGetOrSetNodeInSimpleProgramCache(ref refToWeakReference, refToWeakReferenceIsValid, base.BindStatement(node, diagnostics));
                 }
 
-                BoundNode boundNode = TryGetBoundNodeFromMap(node);
+                // Synthesized statements are not added to the _guardedNodeMap, we cache them explicitly here in  
+                // _lazyGuardedSynthesizedStatementsMap
+                if (statement.WasCompilerGenerated && node.SyntaxTree == _semanticModel.SyntaxTree)
+                {
+                    _semanticModel.GuardedAddSynthesizedStatementToMap(node, statement);
+                }
 
-                if (boundNode == null)
+                return statement;
+            }
+
+            /// <param name="node">Node to use as a key in the cache.</param>
+            /// <param name="boundNode">Not null if there is a cached bound node.</param>
+            /// <param name="resultRefIsValid">Indicates whether returned reference represents a valid reference into the cache, i.e. can be used to cache new bound node in the cache.</param>
+            private unsafe ref WeakReference<BoundNode> TryGetNodeFromSimpleProgramCache<TBoundNode>(SyntaxNode node, out TBoundNode boundNode, out bool resultRefIsValid) where TBoundNode : BoundNode
```

Comment:
I don't see unsafe code in the implementation

Reply:
> I don't see unsafe code in the implementation (I guess use of the Unsafe class may count?)

```AsRef``` uses a pointer for parameter type.

---
In reply to: [384779732](https://github.com/dotnet/roslyn/pull/41926#discussion_r384779732) [](ancestors = 384779732)